R/get-based-on.R

```diff
@@ -0,0 +1,169 @@
+
+#' Get based_on from bbi object
+#'
+#' Returns character vector of the absolute paths all models stored in the `based_on` field of a `bbi_...` object
+#' NOTE: All paths saved in the object or accompanying YAML will be relative **to the location of that YAML**
+#' When the object is loaded into memory, the absolute path to the YAML is stored in the object.
+#' These functions simply stitch together that path with the relative paths from the `based_on` field.
+#' As long as the YAML has not moved since it was loaded, this will work.
+#' @param .bbi_object The model object (or path, etc.) to query
+#' @param .check_exists Logical scaler for whether it will check if the file exists and error if it does not. FALSE by default.
+#' @export
+#' @rdname get_based_on
+get_based_on <- function(.bbi_object, .check_exists = FALSE) {
+  UseMethod("get_based_on")
+}
+
+
+#' The default method attempts to extract the path from any object passed to it
+#' @importFrom fs path_norm
+#' @param .bbi_object The object to attempt to query. Could be a partially built bbi_{.model_type}_object or some other custom object containing model data.
+#' @rdname get_based_on
+#' @export
+get_based_on.default <- function(.bbi_object, .check_exists = FALSE) {
+  tryCatch(
```

Comment:
@dpastoor I worked on refactoring this to just use get_path_from_object under the hood and rediscovered the reasons I didn't do it in the first place. Long story short: I'm gonna leave it like it is, but here are some of the reasons. Number 3 is the main reason.

1. I wanted to die if the key requested is absent, but for `based_on` it is perfectly reasonable for it to be absent, and the function should just return NULL instead of dying.
2. `get_path_from_object` expects the thing it's fetching to be a scaler, not a vector. I can get rid of this, but...
3. the `bbi_run_log_df` dispatch calls `map_chr` so that it can return a character vector of the paths fetched, one per row of the df. _But_ if what's fetched is _not_ a scaler (as in based_on) this fails, which means I need to use `map` instead, but makes the return from that dispatch either a) always a list or b) type-unstable. I don't like either one of those options.

Happy to discuss this further if you want to, but I'd be surprised if you really care. Just wanted to put my thoughts down here so you can follow the thought process. And so we can revisit later if you'd like to.

Reply:
>  I'd be surprised if you really care. 

ouchies ;-)

Yes you are correct and your logic above makes complete sense, esp around map_chr's constrains around scalar returns.